### PR TITLE
EDU 266: How to schedule a Workflow in Go

### PR DESCRIPTION
**Pull request by DJSanti**
## What does this PR do?

This PR adds information on how to work with Schedules in Go.
Includes code blocks. Snippets will be included in a future update.

## Notes to reviewers

Am open to feedback. Please review away.

### CODE REVIEW

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+When used, `list` returns all available Schedules and their Schedule IDs.
+
+To retrieve a list of Schedules, use `ScheduleClient().List()`.
+
+## Pause/unpause
+
+The `pause` and `unpause` actions enable the start or stop of all future Workflow Runs on a given Schedule.
+Pausing a Schedule halts all future Workflow Runs, while unpausing a Schedule allows the Workflow to execute as planned.
+
+Pausing can be enabled upon creating a Schedule by setting `State.Paused` to `true`.
```
Comment (djmagee) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:72]:
```suggestion
Pausing can be enabled when you create a Schedule by setting `State.Paused` to `true`.
```

Alternatively: "When you create a Schedule, you can enable pausing by setting `State.Paused` to `true`."

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+<!--SNIPSTART samples-go-schedule {"selectedLines": ["85"]}-->
+<!--SNIPEND-->
+
+Pausing can also be done by using `Pause()` on the `scheduleHandle`.
```
Comment (djmagee) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:77]:
```suggestion
Pausing can also be done by using `Pause()` on `scheduleHandle`.
```

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+```go
+scheduleHandle.Pause()
+```
+
+To unpause a Schedule, use `Unpause()` on the `scheduleHandle`.
```
Comment (djmagee) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:83]:
```suggestion
To unpause a Schedule, use `Unpause()` on `scheduleHandle`.
```

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+<!--SNIPSTART samples-go-schedule {"selectedLines": ["99-102"]}-->
+<!--SNIPEND-->
+
+## Trigger
+
+Triggering a Schedule immediately executes an Action defined in the Schedule.
+By default, `trigger` is subject to the Overlap Policy.
+
+To trigger a Scheduled Workflow Execution in Go, use `Trigger()` on the `scheduleHandle`.
```
Comment (djmagee) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:93]:
```suggestion
To trigger a Scheduled Workflow Execution in Go, use `Trigger()` on `scheduleHandle`.
```

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 5]
```diff
@@ -0,0 +1,883 @@
...
+<!--SNIPEND-->
+
+## Describe
+
+You can retrieve information about the current Schedule configuration by using the `describe` action.
+This is helpful when you want to get a detailed view of the Schedule and its associated [Workflow Runs](/concepts/what-is-a-run-id).
+
+To describe a Schedule in Go, use `Describe()` on `scheduleHandle`.
+
+<!--SNIPSTART samples-go-schedule {"selectedLines": ["107-110"]}-->
```
[the remark this answers is not in the document]
Comment (DJSanti) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md:280]: Looking at it

**Comment by DJSanti** [comment]
Moving code and stuff over to dacx

**Comment by DJSanti** [comment]
Keeping branch as is. Diagnosing snipsync issue

**Comment by DJSanti** [comment]
Opened PR for smaller snipsyncs: https://github.com/temporalio/samples-go/pull/273

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+tags:
+  - go
+  - developer-guide
+  - how-to
+---
+
+Scheduling [Workflows](/concepts/what-is-a-workflow) is crucial to automation.
+By scheduling [Workflow Executions](/concepts/what-is-a-workflow-execution), you can reduce manual intervention and ensure timely execution of business processes.
+
+Temporal allows you to execute multiple Schedule operations, either through code or with the [CLI tool](/concepts/what-is-the-temporal-cli).
```
Comment (Quinn-With-Two-Ns) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:15]: A user can also use the web UI, might want to call that out.

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+Read on to find out how to create and edit Schedules.
+
+:::note
+Be sure to [enable Schedules in your environment](/concepts/what-is-a-schedule#limitations) before creating any Schedules.
+:::
+
+## Create
+
+Schedules are created with the `create` action.
+For each new Schedule, tbe Temporal Server generates a unique Schedule ID.
```
Comment (Quinn-With-Two-Ns) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:25]: This paragraph says the server generates the schedule ID, that paragraph below says the user must pass an ID. These don't seem consistent

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+<!--SNIPEND-->
+
+## Backfill
+
+Backfilling a Schedule executes [Workflow Tasks](/concepts/what-is-a-workflow-task) ahead of the Schedule's specified time range.
+This is useful when you need to execute a missed or delayed Action, or if you want to test the Workflow ahead of time.
+
+To backfill a Schedule in Go, use `Backfill()` on `scheduleHandle`.
```
Comment (Quinn-With-Two-Ns) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:61]: nit: `scheduleHandle ` -> `ScheduleHandle ` because `ScheduleHandle` is the actual type

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+<!--SNIPSTART samples-go-schedule-create-delete {"selectedLines": ["42-49"]}-->
+
+[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)
+
+```go
+// Delete the schedule once the sample is done
+defer func() {
+	log.Println("Deleting schedule", "ScheduleID", scheduleHandle.GetID())
+	err = scheduleHandle.Delete(ctx)
```
Comment (cretz) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:78]: With these samples showing how to use a `scheduleHandle`, we should show them how to obtain it (using the one from create is often not a realistic use even though we do it in the sample).

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+To unpause a Schedule, use `Unpause()` on `scheduleHandle`.
+
+<!--SNIPSTART samples-go-schedule-unpause-describe {"selectedLines": ["107-110"]}-->
+
+[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)
+
+```go
+// Unpause schedule
+log.Println("Unpausing schedule", "ScheduleID", scheduleHandle.GetID())
+err = scheduleHandle.Unpause(ctx, client.ScheduleUnpauseOptions{})
```
Comment (cretz) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:146]: Inconsistent that unpausing has logs and error checks and everything but pausing is just a single call w/ no error check

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+To trigger a Scheduled Workflow Execution in Go, use `Trigger()` on `scheduleHandle`.
+Set desired changes in `ScheduleTriggerOptions`.
+
+<!--SNIPSTART samples-go-schedule-trigger {"selectedLines": ["55-57"]}-->
+
+[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)
+
+```go
+	// Manually trigger the schedule once
```
Comment (cretz) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:167]: Inconsistent indentation on snippets (some of these are left aligned, some indented)

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+					},
+				},
+				// Run the schedule every 5s
+				Intervals: []client.ScheduleIntervalSpec{
+					{
+						Every: 5 * time.Second,
+					},
+				},
+			}
+			// Start the schedule paused to demonstrate how to unpause a schedule
```
Comment (cretz) [inline docs-src/go/how-to-schedule-a-workflow-in-go.md:218]: This is not starting (may have to go back and fix that comment in samples) and this isn't for demonstrating unpause

File: docs-src/go/how-to-schedule-a-workflow-in-go.md (hunk 7 above)
Comment (DJSanti) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md:25]: Changed per @cretz correction

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+Deletion does not affect any Workflows started by the Schedule.
+
+To delete a Schedule in Go, use the `Delete()` command on `scheduleHandle`.
+
+<!--SNIPSTART samples-go-schedule-create-delete {"selectedLines": ["42-49"]}-->
+
+[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)
+
+```go
+// Delete the schedule once the sample is done
```
[the remark this answers is not in the document]
Comment (DJSanti) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md:75]: Can snip around that

File: docs-src/go/how-to-schedule-a-workflow-in-go.md (hunk 12 above)
Comment (DJSanti) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md:218]: https://github.com/temporalio/samples-go/blob/main/schedule/starter/main.go

File: docs-src/go/how-to-schedule-a-workflow-in-go.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (DJSanti) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md:75]: Again, got it from samples-go: https://github.com/temporalio/samples-go/blob/main/schedule/starter/main.go

File: docs-src/go/how-to-schedule-a-workflow-in-go.md [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+<!--SNIPSTART samples-go-schedule-create-delete {"selectedLines": ["26-38"]}-->
+
+[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)
+
+```go
+// This schedule ID can be user business logic identifier as well.
+scheduleID := "schedule_" + uuid.New()
+workflowID := "schedule_workflow_" + uuid.New()
+// Create the schedule, start with no spec so the schedule will not run.
```
[the remark this answers is not in the document]
Comment (DJSanti) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md:39]: samples-go: https://github.com/temporalio/samples-go/blob/main/schedule/starter/main.go

File: docs-src/go/how-to-schedule-a-workflow-in-go.md (hunk 12 above)
Comment (DJSanti) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md]: There's nothing else demonstrating. I'm working with what's provided, so this looks like a good time to go over the current code samples.

Comment (cretz) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md]: :+1: may have to go back and fix that comment in samples

> There's nothing else demonstrating

Exactly, so should fix that comment

File: docs-src/go/how-to-schedule-a-workflow-in-go.md (hunk 9 above)
Comment (cretz) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md]: > How do we obtain it?

Via `ScheduleClient.GetHandle`

> I am working from what was done in code

You are working with what was done in a single sample that creates and deletes in the same bit of code. But our docs shouldn't expect people would use it that way. Sample code and docs code can often have mismatched purposes. In this case, samples show you how to do everything together at once and docs separate them into standalone sections. These are incompatible in this case.

> If there's another place mentioning the use of scheduleHandle, please let me know

You may have to write it.

Comment (DJSanti) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md]: Very well. I'm going to go ahead and put in code blocks in lieu of the code snippets. Once I migrate this over to dacx format, then I'll write code for a use case.

File: docs-src/go/how-to-schedule-a-workflow-in-go.md (hunk 12 above)
Comment (cretz) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md]: Oh I didn't take it that way, you're good! Ideally we can get the code right from the outset, but if not, ok.

File: docs-src/go/how-to-schedule-a-workflow-in-go.md (hunk 9 above)
Comment (cretz) [reply docs-src/go/how-to-schedule-a-workflow-in-go.md]: May want to confirm with @flossypurse. Maybe updating the sample to break out the steps makes more sense. That's what was done in the Python schedule sample.

File: docs-src/go/how-to-customize-workflow-type-in-go.md [hunk 15]
```diff
@@ -9,7 +9,7 @@ In Go, by default, the Workflow Type name is the same as the function name.
 
 To customize the Workflow Type, set the `Name` parameter with `RegisterOptions` when registering your Workflow with a [Worker](/go/how-to-develop-a-worker-in-go).
 
-<a class="dacx-source-link" href="https://github.com/temporalio/documentation-samples-go/blob/main/yourapp/worker/main_dacx.go">View source code</a>
+<a class="dacx-source-link" href="https://github.com/temporalio/documentation-samples-go/blob/add-go-schedule-sample/yourapp/worker/main_dacx.go">View source code</a>
```
Comment (Quinn-With-Two-Ns) [inline docs-src/go/how-to-customize-workflow-type-in-go.md:12]: Is this intentional?

File: docs-src/go/how-to-develop-a-worker-in-go.md [hunk 16]
```diff
@@ -5,19 +5,6 @@ sidebar_label: Develop Worker
 description: Develop an instance of a Worker by calling worker.New(), available via the go.temporal.io/sdk/worker package.
 ---
 
-The [Worker Process](/concepts/what-is-a-worker-process) is where Workflow Functions and Activity Functions are executed.
-
-- Each [Worker Entity](/concepts/what-is-a-worker-entity) in the Worker Process must register the exact Workflow Types and Activity Types it may execute.
-- Each Worker Entity must also associate itself with exactly one [Task Queue](/concepts/what-is-a-task-queue).
```
Comment (Quinn-With-Two-Ns) [inline docs-src/go/how-to-develop-a-worker-in-go.md:11]: Why remove this? and why in this PR unless maybe I am missing something?

File: docs-src/go/how-to-pause-a-schedule-in-go.md [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+id: how-to-pause-a-schedule-in-go
+title: How to pause a Schedule in Go
+sidebar_label: Pause Schedule
+description: Show how to unpause and pause a Schedule in Go.
+---
+
+`Pause` and `Unpause` enable the start or stop of all future Workflow Runs on a given Schedule.
+
+Pausing a Schedule halts all future Workflow Runs.
+Pausing can be enabled by setting `State.Paused` to `true`, or by using `Pause()` on the ScheduleHandle.
```
Comment (Quinn-With-Two-Ns) [inline docs-src/go/how-to-pause-a-schedule-in-go.md:11]: nit: The word `enabled` normally means the feature can be used. I could see someone reading this getting and confused, thinking they need to change the `State.Paused` setting to allow them to call `Pause()` in the future. I would reword this to how you worded `Unpause()`

**Comment by Quinn-With-Two-Ns** [comment]
Reviewed, I may have missed some files because there appears to be a lot of unrelated changes to other files with nothing to do with schedules?

File: docs-src/go/how-to-customize-workflow-type-in-go.md (hunk 15 above)
Comment (DJSanti) [reply docs-src/go/how-to-customize-workflow-type-in-go.md:12]: In the future, I will hold off on yarn gen until after it gets through a review. But yeah, there will be links to source code before snippets

File: docs-src/go/how-to-describe-a-schedule-in-go.md [hunk 18]
```diff
@@ -0,0 +1,22 @@
+---
+id: how-to-describe-a-schedule-in-go
+title: How to describe a Schedule in Go
+sidebar_label: Describe Schedule
+description: Describe a Schedule in Go.
+---
+
+`Describe` retrieves information about the current Schedule configuration.
+This can include details about the Schedule Spec (such as Intervals), CronExpressions, and Schedule State.
```
[the remark this answers is not in the document]
Comment (DJSanti) [reply docs-src/go/how-to-describe-a-schedule-in-go.md:9]: I can put it in the parentheses, then

[the remark this answers is not in the document]
Comment (DJSanti) [reply docs-src/go/how-to-describe-a-schedule-in-go.md:9]:
```suggestion
This can include details about the Schedule Spec (such as Intervals or, CronExpressions)
```suggestion
This can include details about the Schedule Spec (such as Intervals), CronExpressions, and Schedule State.
``` and Schedule State.
```

File: docs-src/go/how-to-develop-a-worker-in-go.md (hunk 16 above)
Comment (DJSanti) [reply docs-src/go/how-to-develop-a-worker-in-go.md:11]: I just ran yarn gen. This probably got pulled in from somewhere else. Let me check.

**Comment by DJSanti** [comment]
@Quinn-With-Two-Ns I ran yarn gen before I did the review. In the future, I will hold off on that step until an approval is granted.

**Comment by DJSanti** [comment]
@Quinn-With-Two-Ns Also, a lot of these changes occurred because I stated a "ref" branch. Once my changes hit main for the documentation-samples-go repo, this will resolve.

File: docs-src/go/how-to-backfill-a-schedule-in-go.md [hunk 19]
```diff
@@ -0,0 +1,34 @@
+---
+id: how-to-backfill-a-schedule-in-go
+title: How to backfill a Schedule in Go
+sidebar_label: Backfill Schedule
+description: Backfill a Schedule to execute a Workflow at a later time.
+---
+
+Backfilling a [Schedule](/concepts/what-is-a-schedule) executes [Workflow Tasks](/concepts/what-is-a-workflow-task) ahead of the Schedule's specified time range.
```
Comment (Quinn-With-Two-Ns) [inline docs-src/go/how-to-backfill-a-schedule-in-go.md:8]:
```
[Workflow Tasks](/concepts/what-is-a-workflow-task)
```

I think this should be `Action`?

File: docs-src/go/how-to-trigger-a-schedule-in-go.md [hunk 20]
```diff
@@ -0,0 +1,36 @@
+---
+id: how-to-trigger-a-schedule-in-go
+title: How to trigger a Schedule in Go
+sidebar_label: Trigger Schedule
+description: Immediately execute an Action in a Schedule.
+---
+
+Triggering a [Schedule](/concepts/what-is-a-schedule) immediately executes an [Action](/concepts/what-is-a-schedule#action) defined in that Schedule.
+By default, `trigger` is subject to the `AllowAll` Overlap Policy.
```
Comment (dnr) [inline docs-src/go/how-to-trigger-a-schedule-in-go.md:9]: I don't think that's true. Is that default coming from go sdk?

Comment (DJSanti) [reply docs-src/go/how-to-trigger-a-schedule-in-go.md:9]: Yes
